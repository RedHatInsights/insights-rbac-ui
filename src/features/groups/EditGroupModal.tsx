import React, { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { Skeleton } from '@patternfly/react-core/dist/dynamic/components/Skeleton';
import componentTypes from '@data-driven-forms/react-form-renderer/component-types';
import validatorTypes from '@data-driven-forms/react-form-renderer/validator-types';
import componentMapper from '@data-driven-forms/pf4-component-mapper/component-mapper';
import useAppNavigate from '../../hooks/useAppNavigate';
import { ModalFormTemplate } from '../../components/forms/ModalFormTemplate';
import FormRenderer from '../../components/forms/FormRenderer';
import { fetchGroup, updateGroup } from '../../redux/groups/actions';
import { debouncedAsyncValidator } from './validators';

interface Group {
  uuid: string;
  name: string;
  description?: string;
  [key: string]: unknown;
}

interface EditGroupModalProps {
  cancelRoute?: string | { pathname: string; search: string };
  submitRoute?: string | { pathname: string; search: string };
  group?: Group;
  onClose?: () => void; // Optional - used in stories, not in routing
}

export const EditGroupModal: React.FC<EditGroupModalProps> = ({ cancelRoute, submitRoute = cancelRoute, group, onClose }) => {
  const navigate = useAppNavigate();
  const { groupId } = useParams<{ groupId: string }>();
  const dispatch = useDispatch();

  // Get group data and loading state from Redux (like other components do)
  const selectedGroup = useSelector((state: any) => state.groupReducer?.selectedGroup) || group;
  const isLoading = useSelector((state: any) => state.groupReducer?.isRecordLoading);

  // Fetch group data on mount (standard Redux pattern)
  useEffect(() => {
    if (groupId && !group) {
      dispatch(fetchGroup(groupId));
    }
  }, [groupId, group, dispatch]);

  const onSubmit = async (formData: { name: string; description?: string }) => {
    const userData = {
      uuid: selectedGroup?.uuid || '',
      name: formData.name,
      description: formData.description,
    };

    console.log('🚀 Submitting group update:', userData);

    try {
      // Use standard dispatch pattern - redux-promise-middleware handles notifications automatically
      await dispatch(updateGroup(userData));

      // Refresh the group data after successful update
      if (selectedGroup?.uuid) {
        dispatch(fetchGroup(selectedGroup.uuid));
      }

      // Navigate to submitRoute or cancelRoute as fallback
      const route = submitRoute || cancelRoute || '/groups';
      console.log('🔍 Navigating after successful update to:', route);
      navigate(route);
    } catch (error) {
      // Error notifications are handled automatically by redux-promise-middleware
      console.error('Error updating group:', error);
      // Don't navigate on error - let user retry or manually cancel
    }
  };

  // Memoize schema to prevent recreation on every render
  // This is critical for async validators to maintain their debounce state
  const schema = useMemo(
    () => ({
      fields: [
        {
          component: componentTypes.TEXT_FIELD,
          name: 'name',
          type: 'text',
          label: 'Group name',
          placeholder: 'Enter group name',
          isRequired: true,
          autoFocus: true,
          validate: [
            { type: validatorTypes.REQUIRED },
            { type: validatorTypes.MAX_LENGTH, threshold: 150 },
            {
              type: validatorTypes.PATTERN,
              pattern: /^[A-Za-z0-9]+[A-Za-z0-9\s_-]*$/,
              message: 'Group name must start with alphanumeric character and can contain alphanumeric characters, spaces, hyphens, and underscores',
            },
            // Pass current group UUID to validator to exclude it from "already taken" check
            (value: string) => debouncedAsyncValidator(value, 'uuid', selectedGroup?.uuid),
          ],
        },
        {
          component: componentTypes.TEXTAREA,
          name: 'description',
          type: 'text',
          label: 'Description',
          placeholder: 'Enter group description (optional)',
          validate: [{ type: validatorTypes.MAX_LENGTH, threshold: 150 }],
        },
      ],
    }),
    [selectedGroup?.uuid],
  );

  const onCancel = () => {
    // Call onClose if provided (for stories/tests), otherwise just navigate
    if (onClose) {
      onClose();
    } else {
      // Navigate to cancelRoute (used in routing scenarios)
      const route = cancelRoute || '/groups';
      navigate(route);
    }
  };

  // Show loading while data loads (like AddGroupRoles story pattern)
  if (isLoading || !selectedGroup || !selectedGroup.uuid) {
    return (
      <div>
        <Skeleton height="300px" />
      </div>
    );
  }

  return (
    <FormRenderer
      schema={schema}
      subscription={{ values: true, valid: true, pristine: true }}
      FormTemplate={(props: React.ComponentProps<typeof ModalFormTemplate>) => (
        <ModalFormTemplate
          {...props}
          ModalProps={{
            title: `Edit group "${selectedGroup.name}"`,
            isOpen: true,
            variant: 'medium',
            onClose: onCancel,
          }}
          disableSubmit={['validating', 'pristine']}
          submitLabel="Save"
        />
      )}
      initialValues={selectedGroup}
      onSubmit={onSubmit}
      onCancel={onCancel}
      componentMapper={{
        ...componentMapper,
        'text-field': componentMapper[componentTypes.TEXT_FIELD],
        textarea: componentMapper[componentTypes.TEXTAREA],
      }}
    />
  );
};

export default EditGroupModal;
