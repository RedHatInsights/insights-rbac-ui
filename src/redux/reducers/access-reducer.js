import { GET_PRINCIPAL_ACCESS } from '../../redux/action-types';
import { defaultSettings } from '../../helpers/shared/pagination';

// Initial State
export const accessInitialState = {
  isLoading: false,
  access: {
    data: [],
    meta: defaultSettings,
  },
};

const setLoadingState = (state) => ({ ...state, isLoading: true });

const setAccess = (state, { payload }) => ({ ...state, access: payload, isLoading: false });

export default {
  [`${GET_PRINCIPAL_ACCESS}_PENDING`]: setLoadingState,
  [`${GET_PRINCIPAL_ACCESS}_FULFILLED`]: setAccess,
};
