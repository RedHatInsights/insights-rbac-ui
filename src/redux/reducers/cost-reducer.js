import { GET_RESOURCE_DEFINITIONS, GET_RESOURCE } from '../../redux/action-types';
import { defaultSettings } from '../../helpers/shared/pagination';

// Initial State
export const costInitialState = {
  isLoading: false,
  resourceTypes: {
    data: [],
    meta: defaultSettings,
  },
  resources: {},
  loadingResources: 0,
};

const setLoadingState = (state) => ({ ...state, isLoading: true });

const setResourceDefinitions = (state, { payload }) => ({ ...state, resourceTypes: payload, isLoading: false });
const setResource = (state, { payload }) => ({
  ...state,
  resources: { ...state.resources, [payload.links.first.split('/')[5]]: payload.data },
  loadingResources: state.loadingResources - 1,
});

const setResourceLoading = (state) => ({ ...state, loadingResources: state.loadingResources + 1 });

export default {
  [`${GET_RESOURCE_DEFINITIONS}_PENDING`]: setLoadingState,
  [`${GET_RESOURCE_DEFINITIONS}_FULFILLED`]: setResourceDefinitions,
  [`${GET_RESOURCE}_PENDING`]: setResourceLoading,
  [`${GET_RESOURCE}_FULFILLED`]: setResource,
};
