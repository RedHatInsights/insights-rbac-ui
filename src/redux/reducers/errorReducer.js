import { API_ERROR } from '../action-types';

const setErrorState = (_, { payload }) => ({
  errorCode: payload,
});

export default {
  [API_ERROR]: setErrorState,
};
