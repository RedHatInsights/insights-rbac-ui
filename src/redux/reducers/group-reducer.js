import {
  FETCH_GROUP,
  FETCH_SYSTEM_GROUP,
  FETCH_GROUPS,
  RESET_SELECTED_GROUP,
  FETCH_ROLES_FOR_GROUP,
  FETCH_ADD_ROLES_FOR_GROUP,
  FETCH_MEMBERS_FOR_GROUP,
} from '../../redux/action-types';
import omit from 'lodash/omit';
import { defaultSettings } from '../../helpers/shared/pagination';

// Initial State
export const groupsInitialState = {
  groups: {
    data: [],
    meta: defaultSettings,
  },
  selectedGroup: { addRoles: {}, members: { meta: defaultSettings }, pagination: defaultSettings },
  isLoading: false,
  isRecordLoading: false,
};

const setLoadingState = (state) => ({ ...state, isLoading: true });
const setRecordLoadingState = (state) => ({ ...state, isRecordLoading: true, selectedGroup: { ...state.selectedGroup, loaded: false } });
const setRecordRolesLoadingState = (state) => ({ ...state, isRecordRolesLoading: true, selectedGroup: { ...state.selectedGroup, loaded: false } });
const setGroups = (state, { payload }) => ({ ...state, groups: payload, isLoading: false });
const setSystemGroup = (state, { payload }) => ({ ...state, systemGroup: payload?.data?.filter((group) => group?.platform_default)?.[0] });
const setGroup = (state, { payload }) => ({
  ...state,
  isRecordLoading: false,
  groups: {
    ...state.groups,
    data: state.groups.data.map((group) => ({
      ...group,
      ...(payload.uuid === group.uuid && { ...payload, loaded: true }),
    })),
  },
  selectedGroup: {
    ...state.selectedGroup,
    members: { ...state.selectedGroup.members, data: payload.principals },
    ...omit(payload, ['principals', 'roles']),
    loaded: true,
    pagination: { ...state.selectedGroup.pagination, count: payload.roleCount, offset: 0 },
  },
});
const resetSelectedGroup = (state) => ({ ...state, selectedGroup: undefined });
const setRolesForGroup = (state, { payload }) => ({
  ...state,
  isRecordRolesLoading: false,
  selectedGroup: { ...state.selectedGroup, roles: payload.data, pagination: payload.meta, loaded: true },
});

const setMembersForGroupLoading = (state = {}) => ({
  ...state,
  selectedGroup: {
    ...(state.selectedGroup || {}),
    members: { isLoading: true },
  },
});
const setMembersForGroup = (state, { payload }) => ({
  ...state,
  selectedGroup: {
    ...(state.selectedGroup || {}),
    members: {
      isLoading: false,
      ...payload,
    },
  },
});

const setAddRolesLoading = (state) => ({
  ...state,
  selectedGroup: { ...state.selectedGroup, addRoles: { loaded: false } },
});
const setAddRolesForGroup = (state, { payload }) => ({
  ...state,
  selectedGroup: { ...state.selectedGroup, addRoles: { roles: payload.data, pagination: payload.meta, loaded: true } },
});

export default {
  [`${FETCH_GROUPS}_PENDING`]: setLoadingState,
  [`${FETCH_GROUPS}_FULFILLED`]: setGroups,
  [`${FETCH_SYSTEM_GROUP}_PENDING`]: setLoadingState,
  [`${FETCH_SYSTEM_GROUP}_FULFILLED`]: setSystemGroup,
  [`${FETCH_GROUP}_PENDING`]: setRecordLoadingState,
  [`${FETCH_GROUP}_FULFILLED`]: setGroup,
  [`${FETCH_ROLES_FOR_GROUP}_PENDING`]: setRecordRolesLoadingState,
  [`${FETCH_ROLES_FOR_GROUP}_FULFILLED`]: setRolesForGroup,
  [`${FETCH_MEMBERS_FOR_GROUP}_PENDING`]: setMembersForGroupLoading,
  [`${FETCH_MEMBERS_FOR_GROUP}_FULFILLED`]: setMembersForGroup,
  [`${FETCH_ADD_ROLES_FOR_GROUP}_PENDING`]: setAddRolesLoading,
  [`${FETCH_ADD_ROLES_FOR_GROUP}_FULFILLED`]: setAddRolesForGroup,
  [RESET_SELECTED_GROUP]: resetSelectedGroup,
};
