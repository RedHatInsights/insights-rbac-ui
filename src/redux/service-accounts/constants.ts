export const NO_DATA = 'no-data';
export const LAST_PAGE = 'last-page';
export const RESULTS = 'results';
export const ERROR = 'error';
