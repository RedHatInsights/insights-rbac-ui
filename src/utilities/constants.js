export const RBAC_API_BASE = `/api/rbac/v1`;
