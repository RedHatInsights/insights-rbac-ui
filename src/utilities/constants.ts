export const RBAC_API_BASE = `/api/rbac/v1` as const;
export const RBAC_API_BASE_2 = `/api/rbac/v2` as const;
export const COST_API_BASE = `/api/cost-management/v1` as const;
export const INVENTORY_API_BASE = `/api/inventory/v1` as const;

export const DEFAULT_MUA_BUNDLE = 'rhel' as const;

export const DEFAULT_ACCESS_GROUP_ID = 'default-access' as const;
