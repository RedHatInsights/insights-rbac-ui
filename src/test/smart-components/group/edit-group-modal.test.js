import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { shallowToJson } from 'enzyme-to-json';
import { Button } from '@patternfly/react-core';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';
import { RBAC_API_BASE } from '../../../utilities/constants';
import EditGroupModal from '../../../smart-components/group/edit-group-modal';
import { groupsInitialState } from '../../../redux/reducers/group-reducer';
import componentTypes from '@data-driven-forms/react-form-renderer/dist/cjs/component-types';

describe('<EditGroupModal />', () => {
  let initialProps;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;
  let initialState;

  const GroupWrapper = ({ store, children, initialEntries = [] }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      closeUrl: '/groups',
      groupData: {
        name: 'Foo',
        id: '1',
      },
    };
    mockStore = configureStore(middlewares);
    initialState = {
      groupReducer: {
        ...groupsInitialState,
        isLoading: true,
        groupData: {
          name: 'Foo',
          id: '1',
        },
      },
    };
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <EditGroupModal {...initialProps} />
      </Provider>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should redirect back to close URL', (done) => {
    const store = mockStore(initialState);

    apiClientMock.get(
      `${RBAC_API_BASE}/groups/action-modal/1`,
      mockOnce({
        body: {
          component: componentTypes.TEXTAREA_FIELD,
          name: 'comments',
          type: 'text',
          isRequired: false,
          label: 'Comment',
        },
      })
    );

    const wrapper = mount(
      <GroupWrapper store={store} initialEntries={['/groups/edit/:id']}>
        <Route to="/groups/edit/:id" render={(args) => <EditGroupModal {...initialProps} {...args} isOpen />} />
      </GroupWrapper>
    );

    setImmediate(() => {
      wrapper.find(Button).first().simulate('click');
      expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/groups');
      done();
    });
  });
});
