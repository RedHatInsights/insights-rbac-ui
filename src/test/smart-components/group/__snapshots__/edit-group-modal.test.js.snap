// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditGroupModal /> should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Connect(EditGroupModal)
    cancelRoute="/groups"
    filters={Object {}}
    group={
      Object {
        "id": "1",
        "name": "Foo",
      }
    }
    pagination={
      Object {
        "itemCount": 0,
        "limit": 20,
        "offset": 0,
      }
    }
  />
</ContextProvider>
`;
